### DISCUSSION TITLE
onDidCloseTextDocument behavior with Virtual Documents

### DISCUSSION THREAD

**Discussion by Dayveed117** [microsoft/vscode-discussions#47] [Extension Development QnA] [answered]
Is the `onDidCloseTextDocument` event expected to behave abnormally for virtual documents? When an editor of this kind is closed (through UI or shortcut) the event does not trigger. 

I have also noticed that after a period of time that a virtual document editor is not shown (not visible in `visibleTextEditors`), vscode itself closes the document, triggering the event (presumably freeing resources).

Documentation on the event takes note that situations such as this one may occur, suggesting using `onDidChangeVisibleTextEditors`. Is this second event the norm for virtual document handling?

**Comment by isidorn** [comment, marked as answer]
> Documentation on the event takes note that situations such as this one may occur, suggesting using onDidChangeVisibleTextEditors. Is this second event the norm for virtual document handling?

I think yes.
@alexdima can confirm

**Reply by alexdima** [reply]
I confirm that all documents close only after a certain time since all editors showing them are closed, so it is normal that the `onDidCloseTextDocument` event comes some time after closing all editors.
